libglusterfs/src/glusterfs/gf-buf.h

```diff
@@ -0,0 +1,160 @@
+#ifndef _GF_BUF_H_
+#define _GF_BUF_H_
+
+#include "glusterfs/common-utils.h"
+
+#ifndef GF_BUF_STACK_THRESHOLD
+#define GF_BUF_STACK_THRESHOLD 64
+#endif
+
+typedef struct gf_buf {
+    unsigned long heap : 1;
+    unsigned long allocated : __BITS_PER_LONG - 1;
+    unsigned long used;
+    char *data;
+} gf_buf_t;
+
+void
+gf_buf_cleanup(struct gf_buf *);
+
+#define gf_buf_decl_null(var)                                                  \
+    gf_buf_t var __attribute__((__cleanup__(gf_buf_cleanup))) = {0, 0, 0, NULL}
+
+#define gf_buf_decl_empty(var)                                                 \
+    gf_buf_t var __attribute__((__cleanup__(gf_buf_cleanup))) = {0, 0, 1, ""}
+
+#define gf_buf_decl_string(var, str)                                           \
+    gf_buf_t var __attribute__((__cleanup__(gf_buf_cleanup))) = {              \
+        0, 0, strlen(str) + 1, (str)}
```

Comment:
This is a generic buffer. I wouldn't mix it with strings. A buffer only needs to know the available space and the used space. Null termination is a specific string thing.

Once this is ready, a new type `gf_str_t` can be defined to handle the specific particularities of strings.

Reply:
What about a bit indicating a C-style string (buffer contents with null byte terminator)? IOW
```
typedef struct gf_buf {
    unsigned long heap : 1;
    unsigned long string : 1;
    unsigned long allocated : __BITS_PER_LONG - 1;
    unsigned long used : __BITS_PER_LONG - 1;
    char *data;
} gf_buf_t;
```
This way 'gf_buf_concat(&buf0, &buf1)' may either concatenate two strings like 'strcat()' (overwrite null byte at the end of buf0 data) or append data from buf1 at the end of buf0, depend on what buf0 actually is.